### PR TITLE
Changed Vec2

### PR DESCRIPTION
**Pull request by janscas**
1) Allow to operate with scalar values
2) Dynamic class return so it's an object easier to subclass
3) Added \_\_floordiv\_\_ method
4) Added as_type method to return the tuple with a different type (int to float or whatever)

Can this somehow be merged?

Thanks

### CODE REVIEW

File: pyglet/math.py [hunk 1]
```diff
@@ -52,69 +52,263 @@ def clamp(num, min_val, max_val):
...
+        return self.__class__(self[0] + (alpha * (other[0] - self[0])),
+                              self[1] + (alpha * (other[1] - self[1])))
 
     def scale(self, value):
-        return Vec2(self[0] * value, self[1] * value)
+        """Multiply the vector by a scalar value.
+
+        :parameters:
+            `value`  : int or float :
+                The ammount to be scaled by
```
Comment (LennyPhoenix) [inline pyglet/math.py:229]: This looks like a typo here "ammount"

File: pyglet/math.py [hunk 2]
```diff
@@ -52,69 +52,263 @@ def clamp(num, min_val, max_val):
...
+        :rtype: Vec2
+        """
+        return self.__class__(self[0] * value, self[1] * value)
+
+    def rotate(self, angle):
+        """Create a new Vector rotated by the angle. The magnitude remains unchanged.
+
+        :parameters:
+            `angle` : int or float :
+                The angle to rotate by
```
Comment (LennyPhoenix) [inline pyglet/math.py:241]: Is this radians or degrees?

**Review by LennyPhoenix — CHANGES_REQUESTED** [review]
Code looks good to me, just a few little documentation changes imo

File: pyglet/math.py (hunk 1 above)
Comment (janscas) [reply pyglet/math.py:229]: This is not from my code, I just copied from the original Vec2, but I can fix that anyway

File: pyglet/math.py (hunk 2 above)
Comment (janscas) [reply pyglet/math.py:241]: no clue, it's from the original Vec2 code...

Comment (LennyPhoenix) [reply pyglet/math.py:241]: I think it will be radians, might be worth adding that to the docstring

Comment (janscas) [reply pyglet/math.py:241]: i’ll add that

Comment (janscas) [reply pyglet/math.py]: @LennyPhoenix Done!

Comment (LennyPhoenix) [reply pyglet/math.py]: I think "radiants" is a typo, should be all clear once that's fixed!

Comment (janscas) [reply pyglet/math.py]: Done

**Review by LennyPhoenix — APPROVED** [review]
Looks great! Nice work :+1:

**Comment by benmoran56** [comment]
Thanks for the pull request Janscas, and thanks for reviewing it Lenny! 
One small comment: 
It's probably a bit faster to use `Vec2(...` in place of `self.__class__(...` because we can avoid the `self` dot lookup. 
It's a micro-optimization for sure, but it's good if we can keep these classes as fast as possible.

**Comment by janscas** [comment]
> Thanks for the pull request Janscas, and thanks for reviewing it Lenny!
> One small comment:
> It's probably a bit faster to use `Vec2(...` in place of `self.__class__(...` because we can avoid the `self` dot lookup.
> It's a micro-optimization for sure, but it's good if we can keep these classes as fast as possible.

Then keep in mind that the “dot” lookup is extremely fast compared to the call to `isinstance`.
I you want super fast Vec2 then you probably don’t want this feature although I find calling `Vec2(4,5) * 2` ultra useful.

**Comment by benmoran56** [comment]
Sorry for leaving this hanging. 
That's a good point about the `isinstance` cost. Despite this being a useful change, it might not be worth it in that regard.

**Comment by janscas** [comment]
Yep, better close this. Thanks anyway

**Comment by einarf** [comment]
There are still some new methods in this PR that could be useful?

**Comment by benmoran56** [comment]
I believe all of these methods are already existing in the 2.0 branch, so I think we're safe to close it.

**Comment by janscas** [comment]
> I believe all of these methods are already existing in the 2.0 branch, so I think we're safe to close it.

Well, the `__floordiv__` is actually missing from Vec2

**Comment by benmoran56** [comment]
Thanks @janscas, I added that in master.

**Pull request closed without merging**